File: Java/Assignment2/mergeSort.java
/* Implementation of merge sort - a sorting algorithm that divides an array recursively and then merges together the sorted product */
package Assignment2;
import java.util.*;

public class mergeSort {
    public String[] sort(String[] textArray){
        if(textArray.length <= 1){
            return textArray;
        }
        int middle = textArray.length / 2; //find middle of the array

        String[] left = Arrays.copyOfRange(textArray, 0, middle); //grab the first half
        String[] right = Arrays.copyOfRange(textArray, middle, textArray.length); //grab the second half

        left = sort(left); //recursively call sort with left and right halves separately 
        right = sort(right);
        return merge(left, right); //call merge and return solution
    }
    private String[] merge(String[] oneHalf, String[] theOther){

        String[] mergesorted = new String[oneHalf.length + theOther.length]; //create new array the size of the other 2 along with 3 marker vars
        int i = 0;
        int j = 0;
        int k = 0;
        while(i < oneHalf.length && j < theOther.length){ //traverse throgh subarrays
            if(oneHalf[i].compareTo(theOther[j]) <= 0){ //check if string is less than or equal to other. Add string to array and iterate
                mergesorted[k] = oneHalf[i];
                i++;
            }
            else{
                mergesorted[k] = theOther[j];
                j++;
            }
            k++;
        }
        while(i < oneHalf.length){ //if the one is done, but the other is not then continue
            mergesorted[k] = oneHalf[i];
            i++;
            k++;
        }
        while(j < theOther.length){
            mergesorted[k] = theOther[j];
            j++;
            k++;
        }
        return mergesorted;


    }
    
}

File: C++/Assignment2/MergeSort.cpp
//This file creates the sort classes
#include "MergeSort.hpp"

#include <vector>
#include <iostream>
#include <string>

void MergeSort:: mergeSort(vector<string>& magicItems, int start, int end){ //merge sort to sort magic items
    if(start >= end){
        return;
    }
    int middle = (start + end) / 2; //find middle point
    mergeSort(magicItems, start, middle); //sort left
    mergeSort(magicItems, middle + 1, end); //sort right
    merge(magicItems, start, middle, end); //merge sorted arrays
}

void MergeSort:: merge(vector<string>& magicItems, int start, int middle, int end){ //merge sorted arrays together
    //declare left and right pointer. Also create temp sub array of proper length
    int left = start, right = middle + 1;
    vector<string> subVec(end - start + 1);

    //iterate through sub array
    for(int i = 0; i < end - start + 1; i++){
        if(right > end){
            //add element from left side
            subVec[i] = magicItems[left];
            left++;
        }
        else if (left > middle)
        {
            //add element from right side
            subVec[i] = magicItems[right];
            right++;
        }
        else if (magicItems[left].compare(magicItems[right]) < 0)
        {
            //add element from left side
            subVec[i] = magicItems[left];
            left++;
        }
        else{
            //add element from right side
            subVec[i] = magicItems[right];
            right++;
        }
    }
    //move subvector elements to main vector
    for(int j = 0; j < end - start + 1; j++){
        magicItems[start + j] = subVec[j];
    }
}